internal/v4/api.go

```diff
@@ -266,36 +263,39 @@ func (h *Handler) entityExists(id *charm.Reference, req *http.Request) (bool, er
 }
 
 func (h *Handler) baseEntityQuery(id *charm.Reference, selector map[string]int, req *http.Request) (interface{}, error) {
-	var val mongodoc.BaseEntity
-	id1 := *id
-	id1.Series = ""
-	id1.Revision = -1
-	err := h.store.DB.BaseEntities().
-		Find(bson.D{{"_id", &id1}}).
-		Select(selector).
-		One(&val)
-	if err == mgo.ErrNotFound {
+	fields := make([]string, 0, len(selector))
+	for k, v := range selector {
+		if v == 0 {
+			continue
+		}
+		fields = append(fields, k)
+	}
+	val, err := h.store.FindBaseEntity(id, fields...)
+	if errgo.Cause(err) == params.ErrNotFound {
 		return nil, errgo.WithCausef(nil, params.ErrNotFound, "no matching charm or bundle for %s", id)
 	}
 	if err != nil {
 		return nil, errgo.Mask(err)
 	}
-	return &val, nil
+	return val, nil
 }
 
 func (h *Handler) entityQuery(id *charm.Reference, selector map[string]int, req *http.Request) (interface{}, error) {
-	var val mongodoc.Entity
-	err := h.store.DB.Entities().
-		Find(bson.D{{"_id", id}}).
-		Select(selector).
-		One(&val)
-	if err == mgo.ErrNotFound {
+	fields := make([]string, 0, len(selector))
```

Comment:
I think if we did that we'd have to change all of the store methods to be the same, for consistency. I think I prefer a function, although it would only be used here I think.